walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,234 @@
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+
+  /* doEventAndMeetingShareAttendee returns whether the meeting to schedule and a particular event (from the Events Collection) 
+  share attendees. If so, later on, the event must be added to the "bad meeting times" list, since a meeting cannot be scheduled 
+  during the time range of the event.
+   */
+  private boolean doEventAndMeetingShareAttendee(Event event, MeetingRequest meeting, boolean onlyConsiderMandatoryAttendees) {
+    
+    Collection<String> mandatoryAttendeesOfMeeting = meeting.getAttendees();
+    Collection<String> optionalAttendeesOfMeeting = meeting.getOptionalAttendees();
+
+    Set<String> attendeesOfEventSet = event.getAttendees();
+
+    // eventAndMeetingShareAttendee is true if the event and the meeting share at least one attendee
+    boolean eventAndMeetingShareAttendee = false;
+    for (String attendee : attendeesOfEventSet)  
+    { 
+        if (mandatoryAttendeesOfMeeting.contains(attendee)) {
+            eventAndMeetingShareAttendee = true;
+        }
+
+        /* Only check the optional attendees collection for comparison if we should include optional attendees, 
+        as signified by the boolean parameter onlyConsiderMandatoryAttendees */
+        if (!onlyConsiderMandatoryAttendees) {
+             if (optionalAttendeesOfMeeting.contains(attendee)) {
+                eventAndMeetingShareAttendee = true;
+            }
+        }
+
+    } 
+
+    return eventAndMeetingShareAttendee;
+
+  }
+
+  /* Before we find the viable meeting times, we want to sort and reduce the bad meeting times, or in other words: 
+  sort the bad meeting times, then find the overlap in meeting times, then reduce the bad meeting times to larger, non-overlapping 
+  bad meeting times which encompass all of the original bad meeting times. This reduces the work-load of 
+  getViableMeetingDurations later on, and is a neater solution to this problem than just including this code in another function.
+  */
+  private List<TimeRange> sortAndReduce(List<TimeRange> badMeetingTimesParam) {
+
+    List<TimeRange> badMeetingTimes = badMeetingTimesParam;
+    // Sort, so we can find the overlapping times
+    Collections.sort(badMeetingTimes, TimeRange.ORDER_BY_START);
+
+    /* Handle the overlapping events. If events overlap, so long as our comparingIndex is 
+    still in the badMeetingList size range, we want to push our 'furthestBadEnd' (the latest end 
+    time of the overlapping bad meetings) to later times. This handles the case of nested events as well, 
+    since the bad meeting times are sorted by start time from earliest to latest and not end time.
+    */
+    TimeRange firstBadRange;
+
+    int currIndex = 0;
+    int furthestBadStart = 0;
+    int furthestBadEnd = 0;
+
+    // -1 b/c will increment and use currIndex + 1
+    while (currIndex < badMeetingTimes.size() - 1) {
```

Comment:
This definitely works, but the logic is a bit complex and param names are a bit hard to understand.
Not sure if this makes life easier: loop badMeetingTimes, if current meeting and next meeting overlaps, update end time; else add current meeting into mergedBadMeetingTimes (you can rename).